_posts/2024-12-02-safety-of-methods-for-numeric-primitive-types.md

```diff
@@ -0,0 +1,392 @@
+
+In this next code block, we created a macro to generate verification harnesses for each method and type. It showcases the use of macros to generate reusable verification harnesses for different numeric types and methods, ensuring correctness without duplicating code.
+
+```rust
+#[cfg(kani)]
+mod verify {
+    use super::*;
+
+    // Generate harnesses for methods involving two operands
+    // `type`: Integer type
+    // `method`: Method to verify
+    // `harness_name`: A unique identifier for a harness
+    macro_rules! generate_unchecked_math_harness {
+        ($type:ty, $method:ident, $harness_name:ident) => {
+            #[kani::proof_for_contract($type::$method)]
+            pub fn $harness_name() {
+                let num1: $type = kani::any::<$type>(); // (1)
+                let num2: $type = kani::any::<$type>(); // (1)
+                // (2)
+                unsafe { num1.$method(num2); } // (3)
+                // (4)
+            }
+        }
+    }
+
+    generate_unchecked_math_harness!(i8, unchecked_add, unchecked_add_i8);
+    generate_unchecked_math_harness!(i16, unchecked_add, unchecked_add_i16);
+    // ... Repeat for other integer types (32/64/128-bit/architecture-dependent signed/unsigned integers)
+}
+```
+
+A harness contains several parts:
+
+1. [**Symbolic values**](https://model-checking.github.io/kani/tutorial-nondeterministic-variables.html): `num1` and `num2` are symbolic values generated using `kani::any()`. Kani employs symbolic execution to explore a wide range of input possibilities systematically.
+2. **Assumptions**: With a `#[kani::proof_for_contract]` annotation, Kani automatically inserts `kani::assume()` before the unsafe function call. It ensures that all generated values respect the preconditions of `unchecked_add`. If there are additional assumptions not captured by function contracts, you can specify them manually as well. You can find further details in [Kani&#39;s official RFC for function contracts](https://github.com/model-checking/kani/blob/main/rfc/src/rfcs/0009-function-contracts.md#user-experience).
+3. **Unsafe Execution**: The invocation of `unchecked_add` within an unsafe block for verification. [`unsafe`](https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html) code blocks enable unsafe Rust features, such as calling unsafe functions or dereferencing raw pointers.
+4. **Assertions**: Similar to assumptions, with `#[kani::proof_for_contract]` annotation, Kani automatically inserts `kani::assert()` after the unsafe function call. It checks if the function behaves as expected (e.g. returning an expected result) or if certain safety invariants hold after function call.
+
```

Comment:
```suggestion
To summarize: Kani proves that a function's contract holds by assuming preconditions, executing the function under those assumptions, then asserting that the postcondition holds. Kani also checks that the unsafe code does not cause undefined behavior along the way.
```
Feel free to rephrase this--I just think it would be good to have a concluding sentence about what the reader should take away from this section.

Reply:
Thanks! Updated with some minor reformatting